app/repositories/mails.py

```diff
@@ -0,0 +1,182 @@
+    """,
+)
+
+
+async def create(from_id: int, to_id: int, msg: str) -> dict[str, Any]:
+    """Create a new mail entry in the database."""
+    query = f"""\
+        INSERT INTO mail (from_id, to_id, msg, time)
+             VALUES (:from_id, :to_id, :msg, UNIX_TIMESTAMP())
+    """
+    params = {"from_id": from_id, "to_id": to_id, "msg": msg}
+    rec_id = await app.state.services.database.execute(query, params)
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM mail
+         WHERE id = :id
+    """
+    params = {
+        "id": rec_id,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return dict(rec)
+
+
+async def fetch_one(
+    from_id: Optional[int] = None,
+    to_id: Optional[int] = None,
+    time: Optional[str] = None,
+    read: Optional[bool] = None,
+) -> Optional[dict[str, Any]]:
+    """Fetch a mail entry from the database."""
+    if from_id is None and to_id is None and time is None and read is None:
+        raise ValueError("Must provide at least one parameter.")
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM mail
+         WHERE from_id = COALESCE(:from_id, from_id)
+           AND to_id = COALESCE(:to_id, to_id)
+           AND time = COALESCE(:time, time)
+           AND read = COALESCE(:read, read)
+    """
+    params = {
+        "from_id": from_id,
+        "to_id": to_id,
+        "time": time,
+        "read": read,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    return dict(rec) if rec is not None else None
+
+
+async def fetch_count(
+    from_id: Optional[int] = None,
+    to_id: Optional[int] = None,
+    time: Optional[str] = None,
+    read: Optional[bool] = None,
+) -> int:
+    """Fetch the number of mails in the database."""
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM mail
+        WHERE from_id = COALESCE(:from_id, from_id)
+          AND to_id = COALESCE(:to_id, to_id)
+          AND time = COALESCE(:time, time)
+          AND `read` = COALESCE(:read, read)
+    """
+    params = {
+        "from_id": from_id,
+        "to_id": to_id,
+        "time": time,
+        "read": read,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return rec["count"]
+
+
+async def fetch_all(
+    to_id: int,
+    read: bool,
+) -> list[dict[str, Any]]:
+    """Fetch a list of mails from the database."""
+    query = f"""\
+        SELECT m.`msg`, m.`time`, m.`from_id`,
+         (SELECT name FROM users WHERE id = m.`from_id`) AS `from`,
+         (SELECT name FROM users WHERE id = m.`to_id`) AS `to`
+          FROM `mail` m
+         WHERE m.`to_id` = :to_id
+           AND m.`read` = :read
+    """
+    params = {
+        "to_id": to_id,
+        "read": read,
+    }
+
+    recs = await app.state.services.database.fetch_all(query, params)
+    return [dict(rec) for rec in recs]
+
+
+async def update(
```

Comment:
i think this should be modified to only allow the usecase "update messages for these `to` & `from` users to `read=True`, and otherwise mail should be treated strictly as create&read only